tests/unit/s2n_ktls_test.c

```diff
@@ -36,7 +84,115 @@ int main(int argc, char **argv)
+        int fd = 0;
+        EXPECT_OK(s2n_test_configure_ktls_connection(server_conn, &fd));
+
+        server_conn->actual_protocol_version = S2N_TLS13;
+
+        EXPECT_ERROR_WITH_ERRNO(s2n_ktls_validate(server_conn), S2N_ERR_KTLS_UNSUPPORTED_CONN);
+    }
+
+    /* s2n_ktls_validate_socket_mode send */
+    {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        int fd = 0;
+        EXPECT_OK(s2n_test_configure_ktls_connection(server_conn, &fd));
+
+        /* base case */
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_RECV));
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_SEND));
+
+        /* send managed io */
+        server_conn->managed_send_io = false;
+        EXPECT_ERROR_WITH_ERRNO(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_SEND), S2N_ERR_KTLS_SEND_MANAGED_IO);
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_RECV));
+        server_conn->managed_send_io = true;
+
+        /* ktls enabled send */
+        server_conn->ktls_send_enabled = true;
+        EXPECT_ERROR_WITH_ERRNO(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_SEND), S2N_ERR_KTLS_SEND_ENABLED);
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_RECV));
+    }
+
+    /* s2n_ktls_validate_socket_mode recv */
+    {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        int fd = 0;
+        EXPECT_OK(s2n_test_configure_ktls_connection(server_conn, &fd));
+
+        /* base case */
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_RECV));
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_SEND));
+
+        /* recv managed io */
+        server_conn->managed_recv_io = false;
+        EXPECT_ERROR_WITH_ERRNO(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_RECV), S2N_ERR_KTLS_RECV_MANAGED_IO);
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_SEND));
+        server_conn->managed_recv_io = true;
+
+        /* ktls enabled recv */
+        server_conn->ktls_recv_enabled = true;
+        EXPECT_ERROR_WITH_ERRNO(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_RECV), S2N_ERR_KTLS_RECV_ENABLED);
+        EXPECT_OK(s2n_ktls_validate_socket_mode(server_conn, S2N_KTLS_MODE_SEND));
+    }
+
+    /* s2n_ktls_retrieve_file_descriptor */
+    {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        int fd_orig = 1;
+        EXPECT_SUCCESS(s2n_connection_set_write_fd(server_conn, fd_orig));
+
+        int fd_ret = 0;
+        EXPECT_OK(s2n_ktls_retrieve_file_descriptor(server_conn, S2N_KTLS_MODE_SEND, &fd_ret));
+        EXPECT_EQUAL(fd_orig, fd_ret);
+    }
+
```

Comment:
Nit: We now put ";" after test cases to keep clang-format happy (in some cases, but not all, it fails if the ";" isn't there).